lib/widgets/nutrition/charts.dart

```diff
@@ -19,12 +19,128 @@
-import 'package:intl/intl.dart';
 import 'package:wger/helpers/colors.dart';
 import 'package:wger/models/nutrition/nutritional_plan.dart';
 import 'package:wger/models/nutrition/nutritional_values.dart';
 import 'package:wger/widgets/measurements/charts.dart';
 
+class FlNutritionalPlanGoalWidget extends StatefulWidget {
+  const FlNutritionalPlanGoalWidget({
+    super.key,
+    required NutritionalPlan nutritionalPlan,
+  }) : _nutritionalPlan = nutritionalPlan;
+
+  final NutritionalPlan _nutritionalPlan;
+
+  @override
+  State<StatefulWidget> createState() => FlNutritionalPlanGoalWidgetState();
+}
+
+// * fl_chart doesn't support horizontal bar charts yet.
+//   see https://github.com/imaNNeo/fl_chart/issues/113
+//   even if it did, i doubt it would let us put text between the gauges/bars
+// * LinearProgressIndicator has no way to visualize going beyond 100%, or
+//   using multiple colors to show multiple components such as surplus, deficit
+// * here we try drawing our own simple gauges that can go beyond 100%,and we
+//   can use multiple colors...
+class FlNutritionalPlanGoalWidgetState extends State<FlNutritionalPlanGoalWidget> {
+  // normWidth is the width representing 100% completion
+  // note that if val > plan, we will draw beyond this width
+  // therefore, caller must set this width to accommodate surpluses.
+  // why don't we just handle this inside this function? because it might be
+  // *another* gauge that's in surplus and we want to have consistent widths
+  // between all gauges
+  Widget _DIYGauge(BuildContext context, double normWidth, num? plan, double val) {
+    Container segment(double width, Color color) {
+      return Container(
+        height: 16,
+        width: width,
+        decoration: BoxDecoration(color: color, borderRadius: BorderRadius.circular(15.0)),
+      );
+    }
+
+    // paint a simple bar
+    if (plan == null || val == plan) {
+      return segment(normWidth, LIST_OF_COLORS8[0]);
+    }
+
+    // paint a surplus
+    if (val > plan) {
+      return Stack(children: [
+        segment(normWidth * val / plan, COLOR_SURPLUS),
+        segment(normWidth, LIST_OF_COLORS8[0]),
+      ]);
+    }
+
+    // paint a deficit
+    return Stack(children: [
+      segment(normWidth, Theme.of(context).colorScheme.surface),
+      segment(normWidth * val / plan, LIST_OF_COLORS8[0]),
+    ]);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final plan = widget._nutritionalPlan;
+    final today = plan.loggedNutritionalValuesToday;
+
+    return LayoutBuilder(builder: (context, constraints) {
+      var maxVal = 1.0;
+// if any of the bars goes over 100%, find the one that goes over the most
+// that one needs the most horizontal space to show how much it goes over,
+// and therefore reduces the width of "100%" the most, and this width we want
+// to be consistent for all other bars
+      if (plan.goalProtein != null && today.protein / plan.goalProtein! > maxVal) {
```

Comment:
Better use plan.plannedNutritionalValues.protein / fat / etc.

Otherwise we would _only_ use the values from the manual goals, and not the calculated values from the meals

Reply:
good point, but i just noticed an issue with `NutritionalPlan.plannedNutritionalValues`:
* if the planned nutritionalValues come from meals, then a goal of 0 means "this ingredient was not found in any of the mealitems that you've defined, therefore, 0g is your upper limit. if you eat more, you've exceeded your plan.
* if they come from the goals that the user has specified in the nutritional plan settings, than null is translated to 0, and here a goal of 0g means "you've set no restriction on this macro, so you can't hit a limit"

it seems to me this null->0 conversion is a bug. `NutritionalPlan.plannedNutritionalValues` should be able to return null values for goals that are not defined.